torch/_dynamo/backends/cppjit.py

```diff
@@ -0,0 +1,964 @@
+            elif len(op_name) == 2 and op_name == ("tensor", "size"):
+                body.append(
+                    f"auto {node.name} = {node.args[0].name}.size({node.args[1]});"
+                )
+                nodeinfo[node] = NodeInfo(BaseType(BaseTy.int), BaseCType(longT))
+                continue
+
+            # 3. Find the matching NativeFunction.
+            finder = NativeFunctionFinder(nodeinfo)
+            f = finder.find(op_name, node.args, node.kwargs)
+
+            if str(f.func.name) in SCALAR_TO_TENSOR_NATIVEFUNCTION:
+                f = finder.find(SCALAR_TO_TENSOR_NATIVEFUNCTION[str(f.func.name)], node.args, node.kwargs)
+
+            with native_function_manager(f):
+                kernel = Kernel.from_function_and_indices(f, BACKEND_INDICES)
+
+                log.debug(
+                    f"[{node.op}] found function for {build_debug_target_name(node.target)}:"
+                )
+                log.debug(f"""{" " * 4}{f.func}""")
+
+                # 4. Align the given arguments with the function parameters
+                aligned_arguments = align_arguments(
+                    f.func.arguments.flat_all, node.args, node.kwargs
+                )
+                # 5. Match each argument with its desired C++ type
+                #    The types depend on the Signature used (e.g. CppSignature, DispatcherSignature)
+                typed_aligned_arguments = type_aligned_arguments(
+                    aligned_arguments, kernel.sig()
+                )
+
+                # 6. Retrieve the self argument, and generate code for the rest.
+                self_arg = (
+                    f.func.arguments.self_arg.argument
+                    if f.func.arguments.self_arg is not None
+                    else None
+                )
+                arg_values = [
+                    typed_alignedarg_into_cppstr(
+                        a, nodeinfo.get(a.alignedarg.value, None)
+                    )
+                    for a in typed_aligned_arguments
+                    if not is_method or self_arg != a.alignedarg.param
+                ]
+                arg_values_str = ", ".join(str(a) for a in arg_values)
+
+                # 7. Actually generate the function call.
+                if node.op == "call_function":
+                    prefix = f"{kernel.namespace()}::{kernel.name()}"
+                else:
+                    assert node.op == "call_method"
+                    assert self_arg is not None
+                    self_arg_value = next(
+                        filter(lambda a: a.param == self_arg, aligned_arguments)
+                    )
+
+                    assert isinstance(self_arg_value, torch.fx.Node)
+                    prefix = f"{self_arg_value.name}."
+
+                nodeinfo[node] = NodeInfo.from_signature(kernel.sig())
+                body.append(
+                    f"""{nodeinfo[node].cpp_type()} {node.name} = {prefix}({arg_values_str});"""
+                )
```

Comment:
I think you need to identify the tensor arguments which have just been used for the last time and explicitly call `Tensor.reset()` so their memory may be reclaimed. This obviously has a massive impact on memory footprint, but also potentially a significant performance impact because re-used memory may still be in the CPU's cache.

I believe `aot_eager` may also delete inputs, which would allow their memory to be recalimed as well. Certainly that's what inductor's wrapper code does.

Reply:
That makes sense. I will try that.
Thanks, Peter.

Reply:
@ezyang @peterbell10 
I have just tried this, and it worked! 4x speedup! I was not expecting it to be the solution.